### ISSUE TITLE
Add FilterToolbar component as a flexible abstraction around Toolbar (previously DataToolbar) to reduce boilerplate for common filtering use cases

### ISSUE THREAD

**Issue by mturley** [patternfly/patternfly-react#4171]
(tl;dr: I wanted to create a central place to discuss use cases and considerations for a reusable filtering-specific abstraction around the Toolbar component)

While building multiple tables with filtering controls for https://github.com/konveyor/mig-ui based on our current [filterable table demo](https://patternfly-react.surge.sh/documentation/react/demos/filtertabledemo), I found that there is a lot of boilerplate required to get this to work, and that demo is tightly coupled to the specific filter columns/categories in its example. E.g. adding new filters would require new methods equivalent to the `onNameInput` and `onLocationSelect` and new state with some repetition, and the code for building the dropdowns would need to be altered. It is also written as a complex class component, and I didn't want to convert the function components I was working with.

I think this demo still has value as a demonstration of how one might use the building blocks of the Toolbar to make something custom, but I also found myself duplicating a lot of code for the same simple filtering use case in various tables, which became a bit of a mess to maintain. I started experimenting with an abstraction in mig-ui ([draft here](https://github.com/konveyor/mig-ui/tree/master/src/app/common/components/FilterToolbar), needs improvement) which has made it much easier to add different filters to all the other tables in the project.

I started discussing this with @nicolethoen after I saw her discussing different filtering use cases on Slack, and we agreed that it would be nice to make something like this available in PatternFly, as long as we consider most common use cases and make it flexible enough to avoid locking people in an abstraction that won't work for them.

## Basic use case

I have a dropdown of categories (columns), and when a category is selected a filter control is revealed next to it based on the type of category:

* Search filter: a textarea where the user can type a string for a partial value match
* Select filter: a Select component with predetermined options for values to match
* ???: we could enhance this in the future with built-in support for other types of filters
* Custom: we'd export reusable context and props so someone could build e.g. a datepicker filter control, or whatever else they need, and pass it as children.

## Possible API?

The consumer-facing API for the abstraction could look something like this:

```tsx
<FilterToolbar filterValues={filterValues} setFilterValues={setFilterValues}>
  <FilterToolbarSelectControl
    key="location"
    title="Location"
    options={[
      { value: 'raleigh', option: <SelectOption value="Raleigh" /> },
      { value: 'boston', option: <SelectOption value="Boston" /> },
      { value: 'westford', option: <SelectOption value="Westford" /> }
    ]}
    placeholderText="Any"
  />
  <FilterToolbarSelectControl
    key="status"
    title="Status"
    variant={SelectVariant.checkbox}
    options={[
      { value: 'running', option: <SelectOption value="Running" /> },
      { value: 'stopped', option: <SelectOption value="Stopped" /> },
      { value: 'degraded', option: <SelectOption value="Degraded" /> }
    ]}
    placeholderText="Filter by status..."
  />
  <FilterToolbarSearchControl
    key="name"
    title="Name"
    placeholderText="Filter by name..."
  />
</FilterToolbar>
```

And in this example, `filterValues` would be an object that maps category keys to value arrays, e.g.:

```ts
{
	location: ['westford'],
  status: ['stopped', 'degraded'],
  name: ['my-cluster']
}
```

The consumer would be able to store and manage this state and the logic of filtering their data by these values themselves, but we could also provide an optional reusable solution for that ([like I did in mig-ui with a `useFilterState`  React hook here](https://github.com/konveyor/mig-ui/blob/master/src/app/common/duck/hooks.ts#L8-L30)). Chip groups would be handled automatically, and when the user applied a filter, removed a chip or cleared all filters, the `setFilterValues` callback would be called accordingly.

## Other use cases

Nicole mentioned that @dlabrecq was trying to build a filter toolbar with sub-categories (one dropdown reveals another dropdown which reveals the filter control):

<img width="694" alt="Screen Shot 2020-04-21 at 9 38 40 AM" src="https://user-images.githubusercontent.com/811963/80834676-2a205d80-8bbf-11ea-8e1b-4fd02f0a0a02.png">

Also, she mentioned that @jcaianirh was asking about having multiple filter categories in one grouped checkbox select, like [this example](https://patternfly-react.surge.sh/documentation/react/components/select#grouped-checkbox-input-with-filtering). Right now the Toolbar structure assumes a 1-to-1 relationship between the `ToolbarFilter` (filter control wrapper) and a chip group, so we'd need to change it so there can be a 1-to-many relationship there.

These use cases brought up questions about how the existing Toolbar deals with multiple nested ToolbarFilter children, and Nicole and I started brainstorming solutions for that. The general consensus there was that we may need to modify the existing Toolbar to make it more flexible, but we shouldn't complicate it with more functionality to build this abstraction into it, the abstraction should still be its own component and consumers can choose the level of abstraction that meets their needs. The existing component(s) should remain (and maybe be simplified) as building blocks reusable at a lower level.

## Thoughts?

Thanks for reading my lengthy write-up. How do y'all feel about these ideas? I hope to find some time to iterate on my mig-ui prototype of `FilterToolbar` and make a PR for you to play around with soon.

**Comment by mturley** [comment]
@tlabaj @dlabaj @redallen @seanforyou23 I'm curious about your thoughts on this.

**Comment by dlabrecq** [comment]
Additional components is fine, but suspect we will still need to wire up our own custom functionality. We have another (possible) use case where the nested children are up to 5 levels deep.

My biggest issue is that we need a unique `DataToolbarFilter` component associated with each set of options. In addition to a unique `Select` component to display the options. 

In the example above, the tag keys (i.e., 'environment') is associated with 6 options (i.e., tag key values). If we have 10 tag keys, there are 10 `DataToolbarFilter` components. If we have 15 tag keys, we have 15 `DataToolbarFilter` components and so on. 

Ideally, the architecture would allow the `DataToolbarFilter` and select components to be reused with different options.

This code is duplicated for every tag key we display to the user, just populated with different options.

<img width="615" alt="Screen Shot 2020-05-04 at 9 27 04 AM" src="https://user-images.githubusercontent.com/17481322/80970660-849d0200-8de9-11ea-8e0b-c4932325e4c7.png">

**Comment by mturley** [comment]
That makes sense @dlabrecq. I'm imagining a structure where we can have defined categories and subcategories, and it will wire up things for us. Breaking that 1-to-1 relationship between a chip group and a DataToolbarFilter will help I think.

For your use case, maybe there's a component you can pass as a child of `FilterToolbar` that is something like `FilterToolbarSubCategory` with key, title and variant props, and its own children with the actual filter control components. When the component mounts, those children can register themselves with the context provided by FilterToolbar and it can render the DataToolbarFilter and Select components necessary to show the right categories and subcategories defined by its children and grandchildren. I think if we do it right those can be arbitrarily nested however many levels deep you need.

**Comment by dlabrecq** [comment]
Also, in the code above, we display either a `Select` or a `TextInput`. If the tag key values are greater than 50, we ask the user to enter the value rather than attempt to display hundreds of values. 

If we could define the underlying component, in addition to a hierarchy of options that would be helpful.

**Comment by redallen** [comment]
I like the idea of adding your [Filter toolbar components](https://github.com/konveyor/mig-ui/tree/master/src/app/common/components/FilterToolbar) to our library as helpers to Toolbar, but these kinds of helper components haven't really been released in this library before and probably better belong in a demo (at least initially).

While the API does look nicer than our current API, we can't realistically abstract away _every_ possible set of components a consumer might want to use to input filter data to make this the default recommended API.

As for the filter -> many chip group mapping, it sounds like our initial API still needs some work :)

**Comment by mturley** [comment]
> If we could define the underlying component, in addition to a hierarchy of options that would be helpful.

Yep, that's the difference between FilterToolbarSelectControl and FilterToolbarSearchControl. I figured we could make them different instead of using a variant/component prop since they might require different sets of other props. The consumer could also create their own custom control component, we could document an example of that.

**Comment by mturley** [comment]
> I like the idea of adding your [Filter toolbar components](https://github.com/konveyor/mig-ui/tree/master/src/app/common/components/FilterToolbar) to our library as helpers to Toolbar, but these kinds of helper components haven't really been released in this library before and probably better belong in a demo (at least initially).
> 
> While the API does look nicer than our current API, we can't realistically abstract away _every_ possible set of components a consumer might want to use to input filter data to make this the default recommended API.

It's true that we don't have a lot of precedent for this, but as we're starting to provide components that require a lot more complicated usage for the patterns we recommend, it might be nice to offer an (optional) higher level of abstraction. I did that already with FileUpload vs FileUploadField, and our big Table component is a very high level of abstraction (incidentally, for that I think we should also offer the opposite, and export some of the smaller building blocks inside it).

We could simply improve the demo, but I do think this is a huge opportunity to make the toolbar much easier to use. And I agree that we can't abstract away every possible set of filter inputs, I don't think we need to. If we expose the context used by the inputs we do offer, consumers can make their own to plug in. They would still get the benefit of not having to manage all the category selection and chip group management.

**Comment by redallen** [comment]
You make a convincing argument. I'm open to adding them as beta components post-v4 if the documentation surrounding them it is good.

**Comment by mturley** [comment]
Awesome! I'll wait a bit longer to see if anyone else wants to join the discussion, I have plenty of other work to get done. I'm thinking I'll start working on a PR next week or so. I want to nail down a nice extensible structure for allowing custom controls that isn't too verbose.

**Comment by mturley** [comment]
@redallen I wonder if this can be done in two stages with a small change included in v4, since I was thinking we would need to make the existing ToolbarFilter a little more flexible (very simple change, I can make a PR to demonstrate) and that would be the part that is a breaking change.

**Comment by redallen** [comment]
@mturley Sure.

**Comment by nicolethoen** [comment]
I'm here, totally paying attention to and processing this conversation. I've already talked to @mturley about some of the possibilities and definitely will be excited to see the PR. The functionality baked into the Toolbar component (since its name is changing from DataToolbar) was sort of built up around the initial requirement of a 1-to-1 relationship between filter inputs and chip groups. It wasn't until it was built that the FilterableTableDemo was designed and that 1-to-1 relationship was broken that things started to get even crazier. 

So I'm all for reimagining of how multiple, dynamic filters can be managed and the appropriate chip groups dynamically built. It's sort of a reframing of the original question.

Building abstractions on top of the existing toolbar will save consumers some headache and tedious coding, but doesn't solve the original problem. Guess it depends on our priorities at the moment.

**Comment by mturley** [comment]
Yeah! I have a simple fix in mind that will allow us to break that 1-to-1 relationship without much change to the API, but maybe we should revisit the structure in general. When I start on this I'll prioritize a little demo of the 1-to-many relationship and see how that API change feels.

**Comment by mturley** [comment]
@nicolethoen @redallen if you guys want to talk about fixing the original problem with a more thorough reimagining of the structure, don't let me stop you! I was thinking of how we can solve the problem with minimal breakage, but it's beta, so if we want to get that API right maybe breaking it more would be better 😆  I guess that would be a separate issue. Didn't mean to derail the original intention of your discussion @nicolethoen .

**Comment by nicolethoen** [comment]
Nah, you're not derailing anything. I am just not positive to what extent we can address all these increasingly complicated use cases without addressing the API. We are currently twisting a precariously wired component to do a lot more that I imagined it would based on the initial design. The greater the extent to which the abstractions meet consumers' needs, the better.

**Comment by mturley** [comment]
Absolutely. Maybe we can spend some time pair programming or something, and take a look at that API. I wonder what other use cases we might be missing.

**Comment by mturley** [comment]
Still planning to work on this as soon as time allows. Go away stalebot.

**Comment by mturley** [comment]
Still relevant, I have it implemented in two downstream apps and I'm working on abstracting it out in https://github.com/konveyor/lib-ui. Once that is done I will use it as a basis for a PR here.

**Comment by nicolethoen** [comment]
Dont close this

**Comment by nicolethoen** [comment]
@mturley 
This issue has been around for a while - and I think this type of conversation will become very relevant again as we continue to invest in composable solutions.
If you think it's worth contributing an abstraction or wrapper component for a FilterToolbar, I'd be very interested in your thoughts about the best way to present it on the website.

Toolbar is rather old now, so I'm not even sure you're still using it in your product. The conversation about abstraction is worthy having either way, but let me know if it still applies in this case.
